app/scripts/metamask-controller.js

```diff
@@ -1802,6 +1800,27 @@ export default class MetamaskController extends EventEmitter {
     this.emit('update', this.getState())
   }
 
+  /**
+   * @returns {Promise<void>} A promise that resolves when the extension is
+   * unlocked, or immediately if the extension is already unlocked.
+   */
+  getUnlockPromise () {
+    return new Promise((resolve) => {
```

Comment:
Regarding the badge count; maybe we could avoid the complexities you mentioned in the PR description by keeping track of these unlock promises, and counting those in the badge as well.

So instead of returning the promise here, we could add it to a queue that we'd store on the instance:

```
return new Promise((resolve) => {
  if (this.isUnlocked()) {
    resolve()
  } else {
    this.waitingForUnlock.push({ resolve })
  }
})
```

Then on unlock, we could drain the queue:

```
while (this.waitingForUnlock.length >= 0) {
  this.waitingForUnlock.shift().resolve()
}
```

The `updateBadge` function can just count this queue and add that to the badge count.
Though we'd also need to handle updates 🤔 It seems like this'd be better suited to a controller, which we have an established pattern for adding subscriptions to, but I don't know where it'd fit.

Reply:
This was implemented and added to the `AppStateController`. That class now extends `EventEmitter`, and the background subscribes to its `updateBadge` event per the usual pattern.